File: Sheets/src/neetcode/practice/Intervals/_4_meetingRooms.java
package neetcode.practice.Intervals;

import java.util.Comparator;
import java.util.List;

class Interval {
  public int start, end;

  public Interval(int start, int end) {
    this.start = start;
    this.end = end;
  }
}

public class _4_meetingRooms {
  public static boolean canAttendMeetings(List<Interval> intervals) {
    intervals.sort(
        new Comparator<Interval>() {
          @Override
          public int compare(Interval a, Interval b) {
            if (a.end <= b.start) return -1;
            else return 1;
          }
        });
    int n = intervals.size();

    for (int i = 1; i < n; i++) {
      if (intervals.get(i).start < intervals.get(i - 1).end) return false;
    }

    return true;
  }
}

File: Sheets/bin/neetcode/practice/Intervals/_4_meetingRooms.cc
#include "./Interval.hh"

#include <algorithm>
#include <vector>

class Solution {
public:
  inline static auto INTERVAL_ORDER = [](const Interval &a,
                                         const Interval &b) -> bool {
    int cmp = a.start - b.start;
    if (cmp == 0)
      return a.end < b.end;
    else
      return cmp < 0;
  };

  bool canAttendMeetings(std::vector<Interval> &intervals) {
    if (intervals.empty())
      return true;

    std::sort(intervals.begin(), intervals.end(), Solution::INTERVAL_ORDER);
    std::size_t n = intervals.size();

    for (std::size_t i = 0; i < n - 1; i++) {
      Interval current = intervals[i], next = intervals[i + 1];
      if (current.end > next.start)
        return false;
    }

    return true;
  }
};
